File: python/Dave/featureset.py
from Dave.factset import load_factsets, check_schema


def get_filtered(path_or_df, schema=None, date=None, method="snapshot"):
    """get gradient related information for a period of time"""
    if schema is not None and not isinstance(schema, dict):
        raise Exception('Your schema does not have keys, maybe you were trying to set the date parameter')
    schema = check_schema(schema)
    if isinstance(path_or_df, str):
        df = load_factsets(path_or_df, schema)
    else:
        df = path_or_df
    
    if date is not None:
        if isinstance(date, list):
            start_date = min(date)
            end_date = max(date)
            df = df[(df[schema['datetime']] <= end_date) & (df[schema['datetime']] >= start_date)]
        else:
            df = df[df[schema['datetime']] <= date]
    
    if method=="snapshot":
        df = (df.sort_values([schema['entity'], schema['datetime']], 
                  ascending=[False, False])
            .groupby(schema['entity'])
            .first()
            .reset_index()) 
       
    # apply transformations (to do)
    return df
    
    

def get_snapshot(path_or_df, schema=None, date=None):
    """gets the snapshot of the input data at a particular point in time"""    
    return get_filtered(path_or_df, schema, date, method="snapshot")

def get_gradient(path_or_df, schema=None, date=None):
    """get gradient related information for a period of time"""
    return get_filtered(path_or_df, schema, date, method="gradient")
    

File: R/R/featureset.R
#' Gets the gradient related information for a period of time
#'
#' @param df is the path of dataframe object of interest
#' @param schema is the schema of the object
#' @param date is the date which we want to filter on.
#'        This can be either a vector, empty, or single date. If it is a vector
#'        it will filter based on this vector, if it is a single date it will take latest entry based
#'        on the date. If it is NULL it will consider the whole time period
#' @param method is the approach on which the data will be subsetted. Can be either snapshot  or left out (will be assumed to be non-snapshot)
#' @export
get_filtered <- function(df, schema=NULL, date=NULL, method="snapshot") UseMethod("get_filtered")

#' Gets the gradient related information for a period of time
#'
#' @param df is the path of dataframe object of interest
#' @param schema is the schema of the object
#' @param date is the date which we want to filter on.
#'        This can be either a vector, empty, or single date. If it is a vector
#'        it will filter based on this vector, if it is a single date it will take latest entry based
#'        on the date. If it is NULL it will consider the whole time period
#' @param method is the approach on which the data will be subsetted. Can be either snapshot  or left out (will be assumed to be non-snapshot)
#' @seealso\code{\link{get_filtered}}
#' @export
get_filtered.character <- function(df, schema=NULL, date=NULL, method="snapshot") {
  factsets <- load_factsets(df, schema)
  return(get_filtered(factsets, schema, date, method))
}


#' Gets the gradient related information for a period of time
#'
#' @param df is the path of dataframe object of interest
#' @param schema is the schema of the object
#' @param date is the date which we want to filter on.
#'        This can be either a vector, empty, or single date. If it is a vector
#'        it will filter based on this vector, if it is a single date it will take latest entry based
#'        on the date. If it is NULL it will consider the whole time period
#' @param method is the approach on which the data will be subsetted. Can be either snapshot  or left out (will be assumed to be non-snapshot)
#' @seealso\code{\link{get_filtered}}
#' @importFrom dplyr group_by_ filter row_number arrange_ n
#' @export
get_filtered.data.frame <- function(df, schema=NULL, date=NULL, method="snapshot") {
  schema <- check_schema(schema)

  if (!is.null(date)) {
    if(length(date) > 1) {
      start_date <- min(date)
      end_date <- max(date)
      df <- df[df[, schema[['datetime']]] <= end_date & df[, schema[['datetime']]] >= start_date,]
    } else {
      df <- df[df[, schema[['datetime']]] <= date,]
    }
  }

  if (method=="snapshot") {
    return(filter(arrange_(group_by_(df, schema[['entity']]), schema[['datetime']]), row_number() == n()))
  } else {
    return(df)
  }
}

#' Convenience function for get filtered
#'
#' @param df is the path of dataframe object of interest
#' @param schema is the schema of the object
#' @param date is the date which we want to filter on.
#'        This can be either a vector, empty, or single date. If it is a vector
#'        it will filter based on this vector, if it is a single date it will take latest entry based
#'        on the date. If it is NULL it will consider the whole time period
#' @seealso\code{\link{get_filtered}}
#' @export
get_snapshot <- function(df, schema=NULL, date=NULL) {
  return(get_filtered(df, schema, date, method="snapshot"))
}

#' Convenience function for get filtered
#'
#' @param df is the path of dataframe object of interest
#' @param schema is the schema of the object
#' @param date is the date which we want to filter on.
#'        This can be either a vector, empty, or single date. If it is a vector
#'        it will filter based on this vector, if it is a single date it will take latest entry based
#'        on the date. If it is NULL it will consider the whole time period
#' @seealso\code{\link{get_filtered}}
#' @export
get_gradient <- function(df, schema=NULL, date=NULL) {
  return(get_filtered(df, schema, date, method="gradient"))
}



